<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>oliveR: DNA methylation microarray functionalities for sequencing-based data</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">oliveR: DNA methylation microarray
functionalities for sequencing-based data</h1>
<h3 class="subtitle">Andy Madrid</h3>



<div id="general-usage" class="section level2">
<h2>General Usage</h2>
<p>This is a short introduction to the general functionality of the
oliveR package. In short, this package takes a bsseq object, and
performs a plethora of functionalities that were developed for DNA
methlyation microarray data. These include estimating age (epigenetic
age), guessing the sex of samples, identifying lowly-methylated regions
(LMRs), identifiyng stochastic epigenetic mutations (SEMs), identifying
imprinted control regions (ICRs), among other utilities.</p>
<p>Let’s start by installing some packages that are needed to install
and run the madrid package.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">### Install required packages</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="do">### install the oliveR package</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">install_github</span>(<span class="st">&quot;andymadrid/oliveR&quot;</span>)</span></code></pre></div>
<p>Now we can load the oliveR package.</p>
</div>
<div id="setup" class="section level2">
<h2>Setup</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># load in the oliveR pacakge</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(oliveR))</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co">#&gt; No methods found in package &#39;RSQLite&#39; for request: &#39;dbListFields&#39; when loading &#39;lumi&#39;</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;utils::findMatches&#39; by</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; &#39;S4Vectors::findMatches&#39; when loading &#39;ExperimentHubData&#39;</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;GenomicRanges::intersect&#39; by</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; &#39;dplyr::intersect&#39; when loading &#39;oliveR&#39;</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;GenomicRanges::union&#39; by &#39;dplyr::union&#39;</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; when loading &#39;oliveR&#39;</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;bsseq::combine&#39; by &#39;dplyr::combine&#39; when</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; loading &#39;oliveR&#39;</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;GenomicRanges::setdiff&#39; by &#39;dplyr::setdiff&#39;</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt; when loading &#39;oliveR&#39;</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt; Warning: replacing previous import &#39;AnnotationHub::hubUrl&#39; by</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt; &#39;rtracklayer::hubUrl&#39; when loading &#39;oliveR&#39;</span></span></code></pre></div>
<p>Now we can load in a toy dataset.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># get toy example</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">data</span>(bs)</span></code></pre></div>
<p>Let us take a look at this toy example.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># summary of bsseq object</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>bs</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt; An object of type &#39;BSseq&#39; with</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt;   1335479 methylation loci</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt;   10 samples</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; has not been smoothed</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; All assays are in-memory</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co"># summary of phenotypic data associated with bsseq object</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="fu">pData</span>(bs)</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; DataFrame with 10 rows and 3 columns</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt;                 Age   Diagnosis         Sex</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt;           &lt;numeric&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co">#&gt; Sample_1         31     Control      Female</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="co">#&gt; Sample_2         36     Control      Female</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a><span class="co">#&gt; Sample_3         34     Control      Female</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="co">#&gt; Sample_4         53     Control      Female</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a><span class="co">#&gt; Sample_5         59     Control        Male</span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a><span class="co">#&gt; Sample_6         81        Case      Female</span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a><span class="co">#&gt; Sample_7         56        Case        Male</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a><span class="co">#&gt; Sample_8         56        Case      Female</span></span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a><span class="co">#&gt; Sample_9         58        Case        Male</span></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a><span class="co">#&gt; Sample_10        56        Case      Female</span></span></code></pre></div>
<p>We see we have 10 samples. Due to size limitations, this bsseq object
only has the CpGs used to predict age, along with CpGs from the sex
chromosomes. However, for your purposes, there is no need to remove any
CpGs from your bsseq object. In fact, that is not advised.</p>
<p>The bsseq object also has some phenotypic data for the samples, which
includes the age, sex, and (pseudo) diagnostic group for each.</p>
</div>
<div id="guess-sex-of-samples" class="section level2">
<h2>Guess sex of samples</h2>
<p>Now we can attempt to guess the sex of the samples.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># guess sample sex from bsseq object</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>guessed_sex <span class="ot">&lt;-</span> <span class="fu">guessSex</span>(bs)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt; [guessSex]: filtering for only CpGs on sex chromosomes</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt; [guessSex]: getting statistics from sex chromosomes for each sample</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt; [guessSex]: guessing the sex for each sample</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co"># look at the output</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>guessed_sex</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt;           Mean_Cov_chrX Mean_Cov_chrY Sex_Cov_Ratio predictedSex</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; Sample_1       33.20054     0.4870999    0.01467145       Female</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt; Sample_2       36.31087     0.5478362    0.01508739       Female</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt; Sample_3       35.13846     0.4555080    0.01296323       Female</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt; Sample_4       32.87347     0.4562053    0.01387761       Female</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt; Sample_5       15.37184     5.7340866    0.37302542         Male</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt; Sample_6       33.47797     0.4596918    0.01373117       Female</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt; Sample_7       21.84405     7.9071594    0.36198232         Male</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt; Sample_8       36.63454     0.5295641    0.01445532       Female</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt; Sample_9       18.15609     6.7733631    0.37306279         Male</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co">#&gt; Sample_10      33.88088     0.5267607    0.01554743       Female</span></span></code></pre></div>
<p>We see that the guessed sex matches up nicely with the actual sex of
the samples.</p>
</div>
<div id="predict-age" class="section level2">
<h2>Predict age</h2>
<p>Now lets predict the age of the samples using the Methylation Age
Determination via Read-Informed Data (MADRID) madrid() function.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># predict age from bsseq object</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>predicted_age <span class="ot">&lt;-</span> <span class="fu">madrid</span>(bs)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt; [madrid]: will calculate sequencing-based clock (MADRID)</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt; [madrid]: getting methylation levels for each sample</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; [madrid]: filtering CpGs to only those predictive of age</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; [madrid]: no missing data, skipping imputation</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; [madrid]: predicting age now</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; [madrid]: gathering CpGs for array-based clocks now</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt; [madrid]: calculating Dunedin methylation pace now</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt; [madrid]: imputing missing data for array-based CpGs now</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt; [madrid]: calculating age from array-based clocks now</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co"># look at the output</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>predicted_age</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt;           MADRID_Age Horvath_Age Hannum_Age PhenoAge SkinBlood_Age  Lin_Age</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt; Sample_1    34.48592    34.60565 -4.1621911   60.664      8.295322 12.21698</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt; Sample_2    40.80927    34.60565 -3.4333561   60.664      8.665987 12.21698</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt; Sample_3    36.01735    34.60565 -6.1512065   60.664      8.407280 12.21698</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt; Sample_4    54.91187    34.60565 -2.6394282   60.664      8.694256 12.21698</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="co">#&gt; Sample_5    56.54770    34.60565 -0.8991330   60.664     10.024155 12.21698</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="co">#&gt; Sample_6    73.48403    34.60565 -2.3894243   60.664      9.462471 12.21698</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co">#&gt; Sample_7    53.69849    34.60565 -1.8131179   60.664      9.663648 12.21698</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">#&gt; Sample_8    58.76004    34.60565  0.3265283   60.664      8.961952 12.21698</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt; Sample_9    57.31309    34.60565 -1.9617247   60.664      9.503491 12.21698</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="co">#&gt; Sample_10   52.35543    34.60565 -4.1841212   60.664      9.174092 12.21698</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a><span class="co">#&gt;           DunedinPACE</span></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="co">#&gt; Sample_1           NA</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="co">#&gt; Sample_2           NA</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="co">#&gt; Sample_3           NA</span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="co">#&gt; Sample_4           NA</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="co">#&gt; Sample_5           NA</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="co">#&gt; Sample_6           NA</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a><span class="co">#&gt; Sample_7           NA</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a><span class="co">#&gt; Sample_8           NA</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a><span class="co">#&gt; Sample_9           NA</span></span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a><span class="co">#&gt; Sample_10          NA</span></span></code></pre></div>
<p>This outputs the predicted ages for each sample. We can also see how
well they correlate with the actual ages, if that information is readily
available.</p>
<p>Please note, by default the madrid() function will estimate age using
array-based clocks, as well as the sequencing-based clock. To do this,
it leans heavily on two other packages, methylclock, wateRmelon and
DunedinPACE. At its core, for this portion of the function, it really
just formats the methylation matrix accordingly and filters CpGs to only
those on the arrays. In this toy example most of the ages look off due
to the fact that most CpGs are missing because of size limitations. That
being said, in your own data, estimates should (hopefully) be more
accurate.</p>
<p>Please note, if your samples happen to be embryonic, you can set the
option gestationalAge to TRUE. In that case, the function will calculate
the estimated age using several epigenetic clocks trained on prenatal
samples. The estimates will be in “gestational age” for each sample,
which is weeks of gestation.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># assess the correlation between actual and estimated age</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>corMADRID <span class="ot">&lt;-</span> <span class="fu">cor</span>(predicted_age<span class="sc">$</span>MADRID_Age, <span class="fu">pData</span>(bs)<span class="sc">$</span>Age)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>corMADRID</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#&gt; [1] 0.9853733</span></span></code></pre></div>
<div id="test-for-differences" class="section level3">
<h3>Test for differences</h3>
<p>We see that we have a fairly nice correlation between predicted and
actual age. And that’s really all that there is to it. You can take
these results and see if there are any signficiant differences between
the estimated and actual age of samples between groups.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># test for differences (accelerated/decelerated) in actual and estimated age</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>predicted_age<span class="sc">$</span>difference <span class="ot">&lt;-</span> predicted_age<span class="sc">$</span>MADRID_Age <span class="sc">-</span> <span class="fu">pData</span>(bs)<span class="sc">$</span>Age</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">t.test</span>(predicted_age<span class="sc">$</span>difference <span class="sc">~</span> <span class="fu">factor</span>(<span class="fu">pData</span>(bs)<span class="sc">$</span>Diagnosis))</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt;  Welch Two Sample t-test</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt; data:  predicted_age$difference by factor(pData(bs)$Diagnosis)</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; t = -2.0276, df = 7.2851, p-value = 0.08062</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; alternative hypothesis: true difference in means between group Case and group Control is not equal to 0</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt;  -9.1290538  0.6646447</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt; sample estimates:</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt;    mean in group Case mean in group Control </span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co">#&gt;             -2.277782              1.954423</span></span></code></pre></div>
<p>Some people prefer to run the test on the residuals, rather than on
the direct differences in age themselves. That would look something like
this…</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># test for differences in the resduals between actual and estimated age</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>predicted_age<span class="sc">$</span>residuals <span class="ot">&lt;-</span> <span class="fu">resid</span>(<span class="fu">lm</span>(predicted_age<span class="sc">$</span>MADRID_Age <span class="sc">~</span> <span class="fu">pData</span>(bs)<span class="sc">$</span>Age))</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">t.test</span>(predicted_age<span class="sc">$</span>residuals <span class="sc">~</span> <span class="fu">factor</span>(<span class="fu">pData</span>(bs)<span class="sc">$</span>Diagnosis))</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt;  Welch Two Sample t-test</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt; data:  predicted_age$residuals by factor(pData(bs)$Diagnosis)</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt; t = -0.089266, df = 7.1588, p-value = 0.9313</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt; alternative hypothesis: true difference in means between group Case and group Control is not equal to 0</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt;  -3.296327  3.055463</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt; sample estimates:</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt;    mean in group Case mean in group Control </span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt;           -0.06021588            0.06021588</span></span></code></pre></div>
<p>I’d like to note that a Kruskall-Wallis test or something like a
generalized linear model that adjusts for known covariates would
probably be a more approporiate test to use, but for the sake of
demonstration a simple t-test is utilized here.</p>
<p>That being said, in this toy example there is no signficant
difference (accelerated or decelerated) in predicted age between groups.
Hopefully in your data there will be, though!</p>
</div>
<div id="telomere-length" class="section level3">
<h3>Telomere length</h3>
<p>Telomeres are regions of repetitive DNA sequences harbored at the end
of chromosomes. One of the main functions of telomeres is to protect the
terminal ends of chromosomes from degredation. However, one thing to
note is that the length of telomeres shortens with age which can lead to
senescence and apoptosis. As such, shorter telomeres have been linked to
increased risk of diseases in humans. A function has been implemented
here to estimate telomere lengths from sequencing-based data. Let’s take
a look at how this would be done.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># estimate telomere lengths</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>tls <span class="ot">&lt;-</span> <span class="fu">estimateTL</span>(bs)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># look at the output</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>tls</span></code></pre></div>
<p>The above code is not actually run, again, because the toy dataset
we’re working with does not contain the necessary CpGs used to estimate
telomere length. However, in your dataset, that should not be a
problem!</p>
</div>
<div id="estimate-cell-type-proportions-of-blood-samples" class="section level3">
<h3>Estimate cell-type proportions of blood samples</h3>
<p>If you’re using blood or saliva samples, we can also estimate the
cell-type proportions from your sequenced data. The function
estimateCellProps() leans heavily upon the R package meffil to perform
these calculations. In testing, I had some samples that were run on the
EPIC microarray and had the exact same DNA extracts also sequenced and
found that this method produced results that were more correlated with
each other than other methods that were developed for sequence data
(e.g., methylcc).</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># estimate cell-type proportions from blood samples</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>cellProps <span class="ot">&lt;-</span> <span class="fu">estimateCellProps</span>(bs)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co"># check the estimates</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="fu">head</span>(cellProps)</span></code></pre></div>
<p>Please note, the above code is not actually run since the bsseq
object we have here for the demonstration does not have the sufficient
CpGs needed estimate cell-type proportions and would throw out an error.
But, assuming your bsseq object has all the CpGs, that shouldn’t be an
issue for you!</p>
</div>
<div id="convert-hg19-or-hg38-coordinates-to-hg38-or-hg19" class="section level3">
<h3>Convert hg19 (or hg38) coordinates to hg38 (or hg19)</h3>
<p>When working with human data, people often align to one of two genome
builds - either hg19 or hg38. Nowadays, people pretty much just align to
hg38. But, on the off chance that you aligned to hg19 or are using old
data that was aligned to hg19, or if you want to convert your hg38
coordinates to hg19 for whatever reason, I’ve implemented a function to
do just that. See below:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># lift coordinates from hg38 to hg19</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>bsLifted <span class="ot">&lt;-</span> <span class="fu">liftBuild</span>(bs, <span class="at">current =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">new =</span> <span class="st">&quot;hg19&quot;</span>)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#&gt; [liftBuild]: Will lift your bsseq object from hg38 to hg19</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#&gt; loading from cache</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt; [liftBuild]: Lifting over resulted in a loss of 68787 CpGs</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co"># check the lifted coordinates</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>bsLifted</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt; An object of type &#39;BSseq&#39; with</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt;   1266692 methylation loci</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt;   10 samples</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co">#&gt; has not been smoothed</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="co">#&gt; All assays are in-memory</span></span></code></pre></div>
</div>
<div id="imprinted-regions" class="section level3">
<h3>Imprinted regions</h3>
<p>If you happen to have some regions of interest, such as
differentially methylated regions (DMRs), we can also check the overlap
between those segments of the genome and those known to be imprinted.
For more information on how these imprinted regions were identified,
please refer to Jima et al (2022) Epigenetics.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># load in DMR GRanges object</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">data</span>(dmrs)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="co"># take a look at the DMRs</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>dmrs</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="co">#&gt; GRanges object with 476 ranges and 0 metadata columns:</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="co">#&gt;         seqnames              ranges strand</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co">#&gt;            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="co">#&gt;     [1]    chr17   21862802-21863542      *</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="co">#&gt;     [2]     chr2   88778520-88784447      *</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a><span class="co">#&gt;     [3]     chr3   75789914-75795801      *</span></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a><span class="co">#&gt;     [4]     chr3   75786134-75788241      *</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="co">#&gt;     [5]     chr7   61047210-61049745      *</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a><span class="co">#&gt;     ...      ...                 ...    ...</span></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a><span class="co">#&gt;   [472]    chr22   16335697-16336218      *</span></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="co">#&gt;   [473]     chr5   49666198-49666478      *</span></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="co">#&gt;   [474]    chr20   25772827-25773518      *</span></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a><span class="co">#&gt;   [475]    chr10 133688703-133690164      *</span></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a><span class="co">#&gt;   [476]    chr20   30633411-30636228      *</span></span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="co">#&gt;   -------</span></span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a><span class="co">#&gt;   seqinfo: 22 sequences from an unspecified genome; no seqlengths</span></span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a></span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a><span class="co"># overlap with the imprintome</span></span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a>dmrsImprinted <span class="ot">&lt;-</span> <span class="fu">imprintMe</span>(dmrs)</span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a><span class="co">#&gt; [imprintMe]: Looking for overlaps with imprinted regions now</span></span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a></span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a><span class="co"># look at the output</span></span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a>dmrsImprinted</span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a><span class="co">#&gt; GRanges object with 15 ranges and 9 metadata columns:</span></span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a><span class="co">#&gt;        seqnames              ranges strand | seqnames     start       end</span></span>
<span id="cb13-31"><a href="#cb13-31" tabindex="-1"></a><span class="co">#&gt;           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span id="cb13-32"><a href="#cb13-32" tabindex="-1"></a><span class="co">#&gt;    [1]    chr20   29867830-29870156      * |    chr20  29868336  29870377</span></span>
<span id="cb13-33"><a href="#cb13-33" tabindex="-1"></a><span class="co">#&gt;    [2]    chr22   11353466-11355455      * |    chr22  11353621  11355111</span></span>
<span id="cb13-34"><a href="#cb13-34" tabindex="-1"></a><span class="co">#&gt;    [3]     chr9   41272995-41273921      * |    chr9   41273436  41273534</span></span>
<span id="cb13-35"><a href="#cb13-35" tabindex="-1"></a><span class="co">#&gt;    [4]     chr9   63815602-63816764      * |    chr9   63816391  63816562</span></span>
<span id="cb13-36"><a href="#cb13-36" tabindex="-1"></a><span class="co">#&gt;    [5]     chr9   63815602-63816764      * |    chr9   63816764  63816827</span></span>
<span id="cb13-37"><a href="#cb13-37" tabindex="-1"></a><span class="co">#&gt;    ...      ...                 ...    ... .      ...       ...       ...</span></span>
<span id="cb13-38"><a href="#cb13-38" tabindex="-1"></a><span class="co">#&gt;   [11]    chr20   29409184-29410062      * |    chr20  29408910  29409184</span></span>
<span id="cb13-39"><a href="#cb13-39" tabindex="-1"></a><span class="co">#&gt;   [12]     chr7   25864754-25866563      * |    chr7   25866370  25866563</span></span>
<span id="cb13-40"><a href="#cb13-40" tabindex="-1"></a><span class="co">#&gt;   [13]    chr20   30871644-30872386      * |    chr20  30871233  30872165</span></span>
<span id="cb13-41"><a href="#cb13-41" tabindex="-1"></a><span class="co">#&gt;   [14]     chr5 141339607-141341419      * |    chr5  141339675 141339766</span></span>
<span id="cb13-42"><a href="#cb13-42" tabindex="-1"></a><span class="co">#&gt;   [15]    chr10 133688703-133690164      * |    chr10 133689257 133690164</span></span>
<span id="cb13-43"><a href="#cb13-43" tabindex="-1"></a><span class="co">#&gt;            width   strand          ID Parental.Origin.of.Methylation</span></span>
<span id="cb13-44"><a href="#cb13-44" tabindex="-1"></a><span class="co">#&gt;        &lt;integer&gt; &lt;factor&gt; &lt;character&gt;                    &lt;character&gt;</span></span>
<span id="cb13-45"><a href="#cb13-45" tabindex="-1"></a><span class="co">#&gt;    [1]      2042        *    ICR_1174                               </span></span>
<span id="cb13-46"><a href="#cb13-46" tabindex="-1"></a><span class="co">#&gt;    [2]      1491        *    ICR_1337                               </span></span>
<span id="cb13-47"><a href="#cb13-47" tabindex="-1"></a><span class="co">#&gt;    [3]        99        *     ICR_571                               </span></span>
<span id="cb13-48"><a href="#cb13-48" tabindex="-1"></a><span class="co">#&gt;    [4]       172        *     ICR_586                            M**</span></span>
<span id="cb13-49"><a href="#cb13-49" tabindex="-1"></a><span class="co">#&gt;    [5]        64        *    ICR_587^                               </span></span>
<span id="cb13-50"><a href="#cb13-50" tabindex="-1"></a><span class="co">#&gt;    ...       ...      ...         ...                            ...</span></span>
<span id="cb13-51"><a href="#cb13-51" tabindex="-1"></a><span class="co">#&gt;   [11]       275        *    ICR_1162                               </span></span>
<span id="cb13-52"><a href="#cb13-52" tabindex="-1"></a><span class="co">#&gt;   [12]       194        *     ICR_444                               </span></span>
<span id="cb13-53"><a href="#cb13-53" tabindex="-1"></a><span class="co">#&gt;   [13]       933        *    ICR_1186                               </span></span>
<span id="cb13-54"><a href="#cb13-54" tabindex="-1"></a><span class="co">#&gt;   [14]        92        *     ICR_339                            M**</span></span>
<span id="cb13-55"><a href="#cb13-55" tabindex="-1"></a><span class="co">#&gt;   [15]       908        *     ICR_702                               </span></span>
<span id="cb13-56"><a href="#cb13-56" tabindex="-1"></a><span class="co">#&gt;        Nearest.Transcript Distance.to.Nearest.Transcript</span></span>
<span id="cb13-57"><a href="#cb13-57" tabindex="-1"></a><span class="co">#&gt;               &lt;character&gt;                      &lt;numeric&gt;</span></span>
<span id="cb13-58"><a href="#cb13-58" tabindex="-1"></a><span class="co">#&gt;    [1]          RNA5SP532                           3682</span></span>
<span id="cb13-59"><a href="#cb13-59" tabindex="-1"></a><span class="co">#&gt;    [2]            DUX4L41                          70730</span></span>
<span id="cb13-60"><a href="#cb13-60" tabindex="-1"></a><span class="co">#&gt;    [3]           PTGER4P1                           5925</span></span>
<span id="cb13-61"><a href="#cb13-61" tabindex="-1"></a><span class="co">#&gt;    [4]          LINC00537                              0</span></span>
<span id="cb13-62"><a href="#cb13-62" tabindex="-1"></a><span class="co">#&gt;    [5]          LINC00537                              0</span></span>
<span id="cb13-63"><a href="#cb13-63" tabindex="-1"></a><span class="co">#&gt;    ...                ...                            ...</span></span>
<span id="cb13-64"><a href="#cb13-64" tabindex="-1"></a><span class="co">#&gt;   [11]            DUX4L34                           1100</span></span>
<span id="cb13-65"><a href="#cb13-65" tabindex="-1"></a><span class="co">#&gt;   [12]            MIR148A                          83356</span></span>
<span id="cb13-66"><a href="#cb13-66" tabindex="-1"></a><span class="co">#&gt;   [13]            DUX4L38                          22296</span></span>
<span id="cb13-67"><a href="#cb13-67" tabindex="-1"></a><span class="co">#&gt;   [14]          PCDHGA1-2                              0</span></span>
<span id="cb13-68"><a href="#cb13-68" tabindex="-1"></a><span class="co">#&gt;   [15]            DUX4L20                           4065</span></span>
<span id="cb13-69"><a href="#cb13-69" tabindex="-1"></a><span class="co">#&gt;   -------</span></span>
<span id="cb13-70"><a href="#cb13-70" tabindex="-1"></a><span class="co">#&gt;   seqinfo: 22 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>For this to work, your regions should be a GRanges object, with
coordinates stemming from the hg38 version of the human genome. The
output is a GRanges object, with information on the imprinted regions
added to the metadata of your regions that were found to overlap with
them.</p>
<p>Please note, I can’t really make a function to test for enrichment
(e.g., Fisher’s exact test) of your regions in imprinted regions, as
that would require having the background regions from both where your
DMRs/regions came from, and from those used to identify the imprinted
regions - which I do not have, unforunately.</p>
</div>
<div id="identify-icrs" class="section level3">
<h3>Identify ICRs</h3>
<p>While you can overlap your selected regions with those of the
putative ICRs, a new function has been added that allows you to identify
putative ICRs in your own dataset. The function findICRs() looks through
your dataset and tries to identify ICRs which, by default, are regions
with &gt;= 5 consecutive CpGs all with methylation levels ranging from
35-65% in all the CpGs, in all the samples. Let’s test it out and see
how it works.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># identify ICRs in toy dataset</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>icrs <span class="ot">&lt;-</span> <span class="fu">findICRs</span>(bs)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co"># take a look at the ICRs</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>icrs</span></code></pre></div>
<p>The above code is not actually run due to the toy dataset being just
a subste of the genome. As such, we wouldn’t necessarily expect to see
any ICRs, in this case.</p>
</div>
<div id="stochastic-epigenetic-mutations-sems" class="section level3">
<h3>Stochastic epigenetic mutations (SEMs)</h3>
<p>While epigenetic age is the main function of this package, I’ve also
added in a function to identify stochastic epigenetic mutations (SEMs).
For more information on SEMs please refer to Gentilini et al (2015)
Aging and Markov et al (2024) GeroScience.</p>
<p>Essentially, this function takes your data, identifies the
interquartile range (IQR) for each CpG, then sees if there are any
samples (for a given CpG) that has an extremely high (Q3 + 3<em>IQR) or
extremely low (Q1 - 3</em>IQR) methylation level, and adds them all up.
Before you use this function I would like to preface it by saying that
in its current iteration it is extremely slow. Moreover, if you choose
to save the outputs (e.g., the identified SEMs for each individual
sample), it can take up a lot of disk space. So, just fair warning on
that. Either way, here’s how to use it:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># detect SEMs</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="co">#sampleSEMs &lt;- findSEMs(bs, minSamples = 10) # setting samples to 10 for demo, but &gt;50 is recommended for actual use</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="co"># check the output</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>sampleSEMs</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="co">#&gt;           Hyper_SEMs Hypo_SEMs Total_SEMs log10SEMs</span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="co">#&gt; Sample_1         376      5565       5941  3.773860</span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co">#&gt; Sample_2         696      9623      10319  4.013638</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="co">#&gt; Sample_3         512      4501       5013  3.700098</span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="co">#&gt; Sample_4         530      9227       9757  3.989316</span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co">#&gt; Sample_5        6516      5855      12371  4.092405</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="co">#&gt; Sample_6         583      5027       5610  3.748963</span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="co">#&gt; Sample_7        2080      4412       6492  3.812379</span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a><span class="co">#&gt; Sample_8         482      5995       6477  3.811374</span></span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a><span class="co">#&gt; Sample_9        3277      4705       7982  3.902112</span></span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a><span class="co">#&gt; Sample_10        509      9137       9646  3.984347</span></span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a></span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a><span class="co"># check to see if there are any differences between groups</span></span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a><span class="fu">t.test</span>(sampleSEMs<span class="sc">$</span>log10SEMs <span class="sc">~</span> <span class="fu">factor</span>(<span class="fu">pData</span>(bs)<span class="sc">$</span>Diagnosis))</span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-22"><a href="#cb15-22" tabindex="-1"></a><span class="co">#&gt;  Welch Two Sample t-test</span></span>
<span id="cb15-23"><a href="#cb15-23" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb15-24"><a href="#cb15-24" tabindex="-1"></a><span class="co">#&gt; data:  sampleSEMs$log10SEMs by factor(pData(bs)$Diagnosis)</span></span>
<span id="cb15-25"><a href="#cb15-25" tabindex="-1"></a><span class="co">#&gt; t = -0.72429, df = 6.2028, p-value = 0.4953</span></span>
<span id="cb15-26"><a href="#cb15-26" tabindex="-1"></a><span class="co">#&gt; alternative hypothesis: true difference in means between group Case and group Control is not equal to 0</span></span>
<span id="cb15-27"><a href="#cb15-27" tabindex="-1"></a><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span id="cb15-28"><a href="#cb15-28" tabindex="-1"></a><span class="co">#&gt;  -0.2699336  0.1458769</span></span>
<span id="cb15-29"><a href="#cb15-29" tabindex="-1"></a><span class="co">#&gt; sample estimates:</span></span>
<span id="cb15-30"><a href="#cb15-30" tabindex="-1"></a><span class="co">#&gt;    mean in group Case mean in group Control </span></span>
<span id="cb15-31"><a href="#cb15-31" tabindex="-1"></a><span class="co">#&gt;              3.851835              3.913863</span></span></code></pre></div>
<p>You’ll notice that a portion of the script is commented out. That was
done to speed things up. I will be honest and say that this function is
pretty slow. If there’s a future need for it, I can work on speeding it
up. But for now, I’ll leave it as is.</p>
<p>And that’s all there is to it! Similar to above, a Kruskall-Wallis
test or glm would probably be more appropriate for your own data. Just
keep that in mind!</p>
</div>
<div id="lowly-methylated-regions-lmrs" class="section level3">
<h3>Lowly Methylated Regions (LMRs)</h3>
<p>Lowly methylated regions (LMRs) are regions of the genome that
exhibit non-zero, intermediate levels of methylation - often &lt; 50%
methylation. Unlike partially methlyated domains that can span
&gt;100kb, LMRs are more localized. LMRs typically are not located in
CpG islands and, often times, contain transcription factor binding
motifs, suggesting that they can be regulatory elements that define and
maintain cellular identify. Changes in LMRs have been associated in a
plethora of human diseases/disorders/phenomena, such as mutation rate
hetergeneity (Mas-Ponte et al (2024) Nucleic Acids Research), breast
cancer (Fernandez-Jimenez et al (2017) Epigenetics), and embryonic
development (Yizhar-Barnea et al (2018) Scientific Reports). As such, a
function, findLMRs(), has been introduced here to identify LMRs in your
samples. To do this, the function leans on R package MethylSeekR to do
most of the heavy lifting, such as identifying PMDs and training the HMM
to then find LMRs. This is done sample-by-sample. We can take a look at
how one would go about looking for LMRs in their own dataset.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># Identify LMRs</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>lmrs <span class="ot">&lt;-</span> <span class="fu">findLMRs</span>(bs)</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co"># take a look at LMRs</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>lmrs</span></code></pre></div>
<p>Downstream analyses can include finding LMRs that are common amongst
your samples in the same treatment groups, and find those that are
unique to each treatment group.</p>
</div>
<div id="estimate-smoking" class="section level3">
<h3>Estimate smoking</h3>
<p>As was developed for array-based DNA methylation datasets, we can
also estimate smoking status using sequencing-based DNA methylation
data. The function guessSmoking() takes a bsseq object then leans on the
R package wateRmelon to estimate smoking status. By default, the
function guesses “packyears” but can be changed to look at other smoking
measures. See wateRmelon::smokp() for other available
methods/references.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># guess smoking status</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>smoking <span class="ot">&lt;-</span> <span class="fu">guessSmoking</span>(bs)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="co">#&gt; [guessSmoking]: will calculate using Packyears</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="co">#&gt; [guessSmoking]: gathering CpGs for array-based smoking estimation now</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="co">#&gt; [guessSmoking]: estimating smoking status now</span></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="co"># look at the results</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>smoking</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a><span class="co">#&gt;           Packyears</span></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a><span class="co">#&gt; Sample_1   52.64218</span></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a><span class="co">#&gt; Sample_2   52.64218</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a><span class="co">#&gt; Sample_3   52.64218</span></span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a><span class="co">#&gt; Sample_4   52.64218</span></span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a><span class="co">#&gt; Sample_5   52.64218</span></span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a><span class="co">#&gt; Sample_6   52.64218</span></span>
<span id="cb17-16"><a href="#cb17-16" tabindex="-1"></a><span class="co">#&gt; Sample_7   52.64218</span></span>
<span id="cb17-17"><a href="#cb17-17" tabindex="-1"></a><span class="co">#&gt; Sample_8   52.64218</span></span>
<span id="cb17-18"><a href="#cb17-18" tabindex="-1"></a><span class="co">#&gt; Sample_9   52.64218</span></span>
<span id="cb17-19"><a href="#cb17-19" tabindex="-1"></a><span class="co">#&gt; Sample_10  52.64218</span></span></code></pre></div>
<p>As this is just a subset of CpGs in the toy dataset, smoking status
cannot be calculated. But, in your own dataset which should hopefully
have jsut about every CpG in it, the function should be able to estimate
smoking status from your samples.</p>
</div>
<div id="coverage" class="section level3">
<h3>Coverage</h3>
<p>The biggest issue that I have run into while developing this package
has been coverage. For reliable, robust results, higher coverage from
sequencing data is needed. While testing, I found several datasets that
had ~1x genome-wide coverage. As such, estimates for age, sex, LMRs,
SEMs either would highly inaccurate, or were unable to be computed. This
is one of the major drawbacks of sequencing data, relative to the array
where most samples have useable data on just about every CpG tested on
the array. Several imputation methods are utilized in the functions
throuhgout the package, but those can only do so much. So, just be wary
of that fact. As with most sequencing-based analyses: higher coverage
gives better, more robust, and more accurate/reliable results.</p>
</div>
<div id="summary" class="section level3">
<h3>Summary</h3>
<p>Hopefully this package works well for you and your purposes. If there
is some functionality that you would like implemented, please do not
hesitate to email me at madrid2[at]wisc.edu.</p>
<p>Good luck out there!</p>
<p>-AM</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
